pkg/kv/cosmosdb/store.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	var client *azcosmos.Client
+	if params.ReadWriteKey != "" {
+		cred, err := azcosmos.NewKeyCredential(params.ReadWriteKey)
+		if err != nil {
+			return nil, fmt.Errorf("creating key: %w", err)
+		}
+		// Create a CosmosDB client
+		client, err = azcosmos.NewClientWithKey(params.Endpoint, cred, nil)
+		if err != nil {
+			return nil, fmt.Errorf("creating client using access key: %w", err)
+		}
+	} else {
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, fmt.Errorf("default creds: %w", err)
+		}
+		client, err = azcosmos.NewClient("myAccountEndpointURL", cred, nil)
+		if err != nil {
+			return nil, fmt.Errorf("creating client with default creds: %w", err)
+		}
+	}
+
+	// Create container client
+	containerClient, err := client.NewContainer(params.Database, params.Container)
+	if err != nil {
+		return nil, fmt.Errorf("creating container client: %w", err)
+	}
+
+	return &Store{
+		containerClient: containerClient,
+		encoding:        base32.HexEncoding, // Encoding that keeps the strings in-order.
+	}, nil
+}
+
+type Document struct {
+	PartitionKey string `json:"partitionKey"`
+	ID           string `json:"id"`
+	Value        string `json:"value"`
+}
+
+func (s *Store) Get(ctx context.Context, partitionKey, key []byte) (*kv.ValueWithPredicate, error) {
+	if len(partitionKey) == 0 {
+		return nil, kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return nil, kv.ErrMissingKey
+	}
+	item := Document{
+		PartitionKey: s.encoding.EncodeToString(partitionKey),
+		ID:           s.encoding.EncodeToString(key),
+	}
+	pk := azcosmos.NewPartitionKeyString(item.PartitionKey)
+
+	// Read an item
+	itemResponse, err := s.containerClient.ReadItem(ctx, pk, item.ID, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == http.StatusNotFound {
+			return nil, kv.ErrNotFound
+		}
+		return nil, err
+	}
+
+	var itemResponseBody Document
+	err = json.Unmarshal(itemResponse.Value, &itemResponseBody)
+	if err != nil {
+		return nil, err
+	}
+
+	val, err := s.encoding.DecodeString(itemResponseBody.Value)
+	if err != nil {
+		return nil, err
+	}
+	return &kv.ValueWithPredicate{
+		Value:     val,
+		Predicate: kv.Predicate([]byte(itemResponse.ETag)),
+	}, nil
+}
+
+func (s *Store) Set(ctx context.Context, partitionKey, key, value []byte) error {
+	if len(partitionKey) == 0 {
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	// Specifies the value of the partiton key
+	item := Document{
+		PartitionKey: s.encoding.EncodeToString(partitionKey),
+		ID:           s.encoding.EncodeToString(key),
+		Value:        s.encoding.EncodeToString(value),
+	}
+
+	b, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	itemOptions := azcosmos.ItemOptions{
+		ConsistencyLevel: azcosmos.ConsistencyLevelBoundedStaleness.ToPtr(),
```

Comment:
In short, `Strong` is strongly consistent across regions. `BoundedStaleness` is strongly consistent within a single region, but have staleness in other regions. Since we have a kvstore per region, that should be ok.

Long: https://learn.microsoft.com/en-us/azure/cosmos-db/consistency-levels

Reply:
Thanks for the explanation